src/Microsoft.ML.TimeSeries/SrCnnBatchAnomalyDetection.cs

```diff
@@ -0,0 +1,778 @@
+        /// In this mode, output (IsAnomaly, RawScore, Mag, ExpectedValue).
+        /// </summary>
+        AnomalyAndExpectedValue = 2
+    }
+
+    // TODO: SrCnn
+    internal sealed class SrCnnBatchAnomalyDetector : BatchDataViewMapperBase<double, SrCnnBatchAnomalyDetector.Batch>
+    {
+        private const int MinBatchSize = 12;
+        private const int AnomalyOnlyOutputLength = 3;
+        private const int AnomalyAndExpectedValueOutputLength = 4;
+        private const int AnomalyAndMarginOutputLength = 7;
+
+        private readonly int _batchSize;
+        private readonly string _inputColumnName;
+        private readonly int _outputLength;
+        private readonly SrCnnEntireModeler _modler;
+
+        private class Bindings : ColumnBindingsBase
+        {
+            private readonly VectorDataViewType _outputColumnType;
+            private readonly int _inputColumnIndex;
+
+            public Bindings(DataViewSchema input, string inputColumnName, string outputColumnName, VectorDataViewType outputColumnType)
+                : base(input, true, outputColumnName)
+            {
+                _outputColumnType = outputColumnType;
+                _inputColumnIndex = Input[inputColumnName].Index;
+            }
+
+            protected override DataViewType GetColumnTypeCore(int iinfo)
+            {
+                Contracts.Check(iinfo == 0);
+                return _outputColumnType;
+            }
+
+            // Get a predicate for the input columns.
+            public Func<int, bool> GetDependencies(Func<int, bool> predicate)
+            {
+                Contracts.AssertValue(predicate);
+
+                var active = new bool[Input.Count];
+                for (int col = 0; col < ColumnCount; col++)
+                {
+                    if (!predicate(col))
+                        continue;
+
+                    bool isSrc;
+                    int index = MapColumnIndex(out isSrc, col);
+                    if (isSrc)
+                        active[index] = true;
+                    else
+                        active[_inputColumnIndex] = true;
+                }
+
+                return col => 0 <= col && col < active.Length && active[col];
+            }
+        }
+
+        public SrCnnBatchAnomalyDetector(IHostEnvironment env, IDataView input, string inputColumnName, string outputColumnName, double threshold, int batchSize, double sensitivity, SrCnnDetectMode detectMode)
+            : base(env, "SrCnnBatchAnomalyDetector", input)
+        {
+            Contracts.CheckValue(env, nameof(env));
+
+            Contracts.CheckValue(inputColumnName, nameof(inputColumnName));
+            _inputColumnName = inputColumnName;
+
+            env.CheckUserArg(batchSize == -1 || batchSize >= MinBatchSize, nameof(batchSize), "BatchSize must be -1 or no less than 12.");
+            if (batchSize == -1)
+            {
+                _batchSize = (int)input.GetRowCount();
+            }
```

Comment:
In the transformer version, we allow batchSize to be -1, means using the whole data set as one batch. Here I try to get row count from input when batchSize=-1 and set an actual positive batchSize. But I found that if we load dataview by text file loader, we could not get this count by this way. Is there any other way to achieve this?

Reply:
ML.NET does lazy evaluation of most things. So the textloader may not know the number of rows ahead of time. When batchSize is -1, you will be computing the anomaly on the whole file and so you are going to read the whole file before doing your computation. You will need to pass the -1 value of _batchSize further and calculate the actual row count after you have read all the data and before computation begins.

---
In reply to: [425803839](https://github.com/dotnet/machinelearning/pull/5135#discussion_r425803839) [](ancestors = 425803839)

Reply:
Correct. I commented the same thing above before I noticed your comment. 
If you need the number of rows in the data, you can compute it here in the ctor, by getting a row cursor with no active columns and counting how many times `MoveNext()` returns true. So you can do something like this:
```
var rowCount = input.GetRowCount();
if (rowCount!=null)
    _batchSize = rowCount;
else
{
    _batchSize = 0;
    using (var cursor = input.GetRowCursor())
    {
        while (cursor.MoveNext())
            _batchSize++;
    }
}
```

---
In reply to: [426029554](https://github.com/dotnet/machinelearning/pull/5135#discussion_r426029554) [](ancestors = 426029554,425803839)

Reply:
By the way, you would need to check for overflow in this case, since the row count can be greater than `int.MaxValue`.

---
In reply to: [426407146](https://github.com/dotnet/machinelearning/pull/5135#discussion_r426407146) [](ancestors = 426407146,426029554,425803839)

Reply:
I have some concern about count batchSize here since this may introduce an extra traversing. Could we allow batchSize=-1 when initialize Batch object, and in GetIsNewBatchDelegate and GetLastInBatchDelegate, add a special condition for _batchSize=1? In Batch class, since data will be read in later, there is no need to get batchSize from outside when no batch is actually used.

---
In reply to: [426407783](https://github.com/dotnet/machinelearning/pull/5135#discussion_r426407783) [](ancestors = 426407783,426407146,426029554,425803839)

Reply:
Fixed, please check.

---
In reply to: [426719317](https://github.com/dotnet/machinelearning/pull/5135#discussion_r426719317) [](ancestors = 426719317,426407783,426407146,426029554,425803839)